### PR TITLE
Feature/nicoapg/18 implement a version of toxgan cool project from the fda

### PR DESCRIPTION
**Pull request by nicoapg**
Added functionality based on work available online, adapted code and made it functional as per in DEMO file. 

- Dummy data is made available as per in original repository
- Added models placeholder
- Added a client for GAN generation
- Added config with relevant variables
- Added launch demo tox pipeline file to summarize code

### CODE REVIEW

File: vianu/toxicologygan/src/client.py [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+        D = scaler.fit_transform(np.array(D, dtype=np.float32).reshape(len(D), -1))
+        D = torch.tensor(D, device=device)
+        dataset = torch.utils.data.TensorDataset(M, S, T, D)
+        dataloader = torch.utils.data.DataLoader(
+            dataset, batch_size=batch_size, shuffle=True
+        )
+
+        return dataloader
+
+    def Time(self, SACRIFICE_PERIOD):
```
Comment (jaegglic) [inline vianu/toxicologygan/src/client.py:310]: static method?

File: vianu/toxicologygan/src/client.py [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+        switcher = {
+            "4 day": 4 / 29,
+            "8 day": 8 / 29,
+            "15 day": 15 / 29,
+            "29 day": 29 / 29,
+        }
+        return switcher.get(SACRIFICE_PERIOD, "error")
+
+    def Dose(self, DOSE_LEVEL):
```
Comment (jaegglic) [inline vianu/toxicologygan/src/client.py:319]: static method?

File: vianu/toxicologygan/launch_demo_tox_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+# USE TRAINED MODEL TO GENERATE NEW DATA
+path = working_dir
+
+# Read data
+treatments = pd.read_csv(
+    os.path.join(path, "dummy_data", "Example_Treatments_test.tsv"), sep="\t"
+)
+training_data = pd.read_csv(
+    os.path.join(path, "dummy_data", "Example_Data_training.tsv"), sep="\t"
```
Comment (jaegglic) [inline vianu/toxicologygan/launch_demo_tox_pipeline.py:79]: minor remark: as path (being working_path from config.py) is a `pathlib.Path` object, you don't need to use os.path.join but can simply use
`model_path = path / 'dummy_data' / 'Example_Data_training.tsv''`

Also for the following ones you could use the overloaded /-operator from `pathlib.Path`

**Review by jaegglic — APPROVED** [review]
I have some generic remarks:
- can we use type-def for all function arguments and return values? I think this helps a lot in understanding what the function does or needs.
- PEP8 hase some guidelines on naming convention (lower_case for variables, arguments and function names, UPPER_CASE for constants and CamelCase for class definitions) -> https://peps.python.org/pep-0008/#prescriptive-naming-conventions. I think it helps for improving readability

File: docs/toxicologygan.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
+---
+title: ToxicologyGAN
+layout: home
+nav_order: 3
+---
+
+# Welcome to ToxicologyGAN!
+
+**ToxicologyGAN** is a *Gradio-based* application inspired on the work of Chen et al. (2023) (A generative adversarial 
+network model alternative to animal studies for clinical pathology assessment) designed to help  users to.... TODO
```
Comment (alho94) [inline docs/toxicologygan.md:10]: can we replace the TODO with content already? I feel if not, we might end up forgetting it at al.

File: docs/toxicologygan.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+2. **STEP 2**:
+    - **Pipeline_1**: Uses X model to predict...
+    - **Pipeline_2**: Uses X Model to infer...
+
+3. **STEP 3**:
+    - **FEATURE 1** ...
+    - **FEATURE 2** ...
+    - **FEATURE 3** ...
+
```
Comment (alho94) [inline docs/toxicologygan.md:28]: same as previously - should documentation be the same PR as code?

File: vianu/toxicologygan/src/client.py [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+        gen_input = torch.cat([noise, Stru, Time, Dose], -1)
+        Measurement = self.model(gen_input)
+        return Measurement
+
+
+class Discriminator(nn.Module):
+    def __init__(self, Stru_dim, Time_dim, Dose_dim, Measurement_dim):
+        super(Discriminator, self).__init__()
+
+        self.model = nn.Sequential(
```
Comment (alho94) [inline vianu/toxicologygan/src/client.py:451]: just out of curiosity, why this architecture?

File: vianu/toxicologygan/src/client.py [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+
+logger = logging.getLogger("toxicologygan_logger")
+
+
+class ToxicologyGANClient:
+    """
+    The client that orchestrates the data fetching and training.
+    """
+
+    def __init__(self):
```
[the remark this answers is not in the document]
Comment (nicoapg) [reply vianu/toxicologygan/src/client.py:19]: init? sorry I did not understand the comment

File: vianu/toxicologygan/src/train.py
Comment (alho94) [inline vianu/toxicologygan/src/train.py:1]: delete?

**Review by alho94 — APPROVED** [review]
looks good, just some questions =)

File: vianu/toxicologygan/src/client.py [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+            pd.DataFrame(
+                scaler.fit_transform(data.iloc[:, 3:]), columns=data.columns[3:]
+            )
+        )
+
+        S = pd.DataFrame(columns=descriptors.columns)
+        M = pd.DataFrame(columns=data.columns[3:])
+        T = []
+        D = []
+        for i in range(len(data)):
```
[the remark this answers is not in the document]
Comment (nicoapg) [reply vianu/toxicologygan/src/client.py:286]: Nice! I am adapting it

File: vianu/toxicologygan/src/client.py [hunk 9]
```diff
@@ -0,0 +1,473 @@
...
+                    optimizer_G.step()
+                print(
+                    "[Epoch %d/%d] [Batch %d/%d] [D loss: %f] [G loss: %f]"
+                    % (
+                        epoch + 1,
+                        n_epochs,
+                        i + 1,
+                        len(dataloader),
+                        d_loss.item(),
+                        g_loss.item(),
```
[the remark this answers is not in the document]
Comment (nicoapg) [reply vianu/toxicologygan/src/client.py:233]: yes I get your point, g_loss gets defined immediately after a single batch is used, but the printing is meant to report statistics every time a batch is used. I think we should leave it as is.

File: vianu/toxicologygan/src/client.py [hunk 10]
```diff
@@ -0,0 +1,473 @@
...
+                    generator.state_dict(),
+                    os.path.join(model_path, "generator_{}".format(epoch + 1)),
+                )  # nosec
+
+    def create_custom_dataloader(
+        self,
+        data_filepath: str,
+        descriptors_path: str,
+        batch_size: int,
+        device: Union[str, torch.device],
```
[the remark this answers is not in the document]
Comment (nicoapg) [reply vianu/toxicologygan/src/client.py:251]: not sure I know which one you mean. Shuffle came into the training data as a boolean value

File: vianu/toxicologygan/src/client.py (hunk 6 above)
Comment (nicoapg) [reply vianu/toxicologygan/src/client.py:481]: we can talk in the office :)

File: docs/toxicologygan.md (hunk 5 above)
Comment (nicoapg) [reply docs/toxicologygan.md:28]: separated the docs

File: docs/toxicologygan.md (hunk 4 above)
Comment (nicoapg) [reply docs/toxicologygan.md:10]: same same -> moved to another PR

**Review by nicoapg — COMMENTED** [review]
Updated all requested changes
